File: mini-agent-harness/typescript/agent/extensionDecision.ts
import type {
  PermissionDecision,
  PermissionGate,
  PermissionRequest,
} from './permissions.ts'

export type DecisionBehavior = 'allow' | 'ask' | 'deny'

export type DecisionEvidence = Readonly<{
  stage: 'pre-hook' | 'policy' | 'resolver' | 'post-hook' | 'abort'
  sourceId: string
  behavior: DecisionBehavior | 'continue' | 'stop'
  revision: number
}>

export type DecisionContext = Readonly<{
  callId: string
  toolName: string
  revision: number
  input: Readonly<Record<string, unknown>>
}>

export type PreHookResult = Readonly<{
  behavior?: DecisionBehavior
  updatedInput?: Readonly<Record<string, unknown>>
}>

export type PreDecisionHook = Readonly<{
  id: string
  run(context: DecisionContext, signal: AbortSignal): PreHookResult | Promise<PreHookResult>
}>

export type PostDecisionHook = Readonly<{
  id: string
  run(
    context: DecisionContext & Readonly<{ succeeded: boolean }>,
    signal: AbortSignal,
  ): Readonly<{ blockContinuation?: boolean }> | Promise<Readonly<{ blockContinuation?: boolean }>>
}>

export type AskResolution = PermissionDecision & Readonly<{
  updatedInput?: Readonly<Record<string, unknown>>
}>

export type AskResolver = (
  context: DecisionContext,
  evidence: readonly DecisionEvidence[],
  signal: AbortSignal,
) => AskResolution | Promise<AskResolution>

export type PreparedDecision = Readonly<{
  context: DecisionContext
  decision: PermissionDecision
  evidence: readonly DecisionEvidence[]
}>

export type PostDecision = Readonly<{
  continueConversation: boolean
  evidence: readonly DecisionEvidence[]
}>

export class ExtensionDecisionPipeline {
  readonly #preHooks: readonly PreDecisionHook[]
  readonly #postHooks: readonly PostDecisionHook[]
  readonly #askResolver?: AskResolver

  constructor(options: Readonly<{
    preHooks?: readonly PreDecisionHook[]
    postHooks?: readonly PostDecisionHook[]
    askResolver?: AskResolver
  }> = {}) {
    this.#preHooks = Object.freeze([...(options.preHooks ?? [])])
    this.#postHooks = Object.freeze([...(options.postHooks ?? [])])
    this.#askResolver = options.askResolver
    assertUniqueHookIds([...this.#preHooks, ...this.#postHooks])
  }

  async prepare(options: Readonly<{
    callId: string
    toolName: string
    input: Readonly<Record<string, unknown>>
    validateSchema(input: Readonly<Record<string, unknown>>): void
    validateSemantics?(input: Readonly<Record<string, unknown>>): void | Promise<void>
    permissionRequest(input: Readonly<Record<string, unknown>>): PermissionRequest
    gate: PermissionGate
    signal: AbortSignal
  }>): Promise<PreparedDecision> {
    throwIfAborted(options.signal)
    let context = freezeContext(options.callId, options.toolName, 0, options.input)
    const evidence: DecisionEvidence[] = []
    const proposals: DecisionBehavior[] = []
    options.validateSchema(context.input)
    if (options.validateSemantics) {
      await options.validateSemantics(context.input)
    }

    for (const hook of this.#preHooks) {
      throwIfAborted(options.signal)
      const result = Object.freeze(await hook.run(context, options.signal))
      if (result.updatedInput !== undefined) {
        context = freezeContext(
          context.callId,
          context.toolName,
          context.revision + 1,
          result.updatedInput,
        )
        options.validateSchema(context.input)
        if (options.validateSemantics) {
          await options.validateSemantics(context.input)
        }
      }
      if (result.behavior) {
        proposals.push(result.behavior)
        evidence.push(freezeEvidence('pre-hook', hook.id, result.behavior, context.revision))
      }
    }

    throwIfAborted(options.signal)
    const policy = Object.freeze(await abortable(
      options.gate.decide(options.permissionRequest(context.input), options.signal),
      options.signal,
    ))
    const policyBehavior: DecisionBehavior = policy.allowed ? 'allow' : 'deny'
    evidence.push(freezeEvidence('policy', 'permission-gate', policyBehavior, context.revision))

    if (!policy.allowed || proposals.includes('deny')) {
      const reason = !policy.allowed ? policy.reason : 'pre-hook-deny'
      return freezePrepared(context, { allowed: false, reason }, evidence)
    }

    if (proposals.includes('ask')) {
      if (!this.#askResolver) {
        evidence.push(freezeEvidence('resolver', 'missing-resolver', 'deny', context.revision))
        return freezePrepared(
          context,
          { allowed: false, reason: 'approval-required-without-resolver' },
          evidence,
        )
      }
      let resolved: AskResolution
      try {
        resolved = Object.freeze(await abortable(
          this.#askResolver(context, Object.freeze([...evidence]), options.signal),
          options.signal,
        ))
      } catch (error) {
        if (options.signal.aborted) throw error
        evidence.push(freezeEvidence('resolver', 'resolver-error', 'deny', context.revision))
        return freezePrepared(context, { allowed: false, reason: 'approval-resolver-failed' }, evidence)
      }
      evidence.push(freezeEvidence(
        'resolver',
        'ask-resolver',
        resolved.allowed ? 'allow' : 'deny',
        context.revision,
      ))
      if (!resolved.allowed) return freezePrepared(context, resolved, evidence)
      if (resolved.updatedInput !== undefined) {
        context = freezeContext(
          context.callId,
          context.toolName,
          context.revision + 1,
          resolved.updatedInput,
        )
        options.validateSchema(context.input)
        if (options.validateSemantics) {
          await options.validateSemantics(context.input)
        }
        throwIfAborted(options.signal)
        const rewrittenPolicy = Object.freeze(await abortable(
          options.gate.decide(options.permissionRequest(context.input), options.signal),
          options.signal,
        ))
        evidence.push(freezeEvidence(
          'policy',
          'permission-gate-after-resolver-rewrite',
          rewrittenPolicy.allowed ? 'allow' : 'deny',
          context.revision,
        ))
        if (!rewrittenPolicy.allowed) {
          return freezePrepared(context, rewrittenPolicy, evidence)
        }
      }
    }

    throwIfAborted(options.signal)
    return freezePrepared(context, policy, evidence)
  }

  async after(
    prepared: PreparedDecision,
    succeeded: boolean,
    signal: AbortSignal,
  ): Promise<PostDecision> {
    const evidence = [...prepared.evidence]
    let continueConversation = true
    for (const hook of this.#postHooks) {
      const result = Object.freeze(await hook.run(
        Object.freeze({ ...prepared.context, succeeded }),
        signal,
      ))
      if (result.blockContinuation) {
        continueConversation = false
        evidence.push(freezeEvidence(
          'post-hook',
          hook.id,
          'stop',
          prepared.context.revision,
        ))
      } else {
        evidence.push(freezeEvidence(
          'post-hook',
          hook.id,
          'continue',
          prepared.context.revision,
        ))
      }
    }
    return Object.freeze({
      continueConversation,
      evidence: Object.freeze(evidence),
    })
  }
}

function freezeContext(
  callId: string,
  toolName: string,
  revision: number,
  input: Readonly<Record<string, unknown>>,
): DecisionContext {
  return Object.freeze({
    callId,
    toolName,
    revision,
    input: deepFreeze(structuredClone(input)),
  })
}

function freezeEvidence(
  stage: DecisionEvidence['stage'],
  sourceId: string,
  behavior: DecisionEvidence['behavior'],
  revision: number,
): DecisionEvidence {
  return Object.freeze({ stage, sourceId, behavior, revision })
}

function freezePrepared(
  context: DecisionContext,
  decision: PermissionDecision,
  evidence: readonly DecisionEvidence[],
): PreparedDecision {
  return Object.freeze({
    context,
    decision: Object.freeze({ ...decision }),
    evidence: Object.freeze([...evidence]),
  })
}

function deepFreeze<T>(value: T): T {
  if (value && typeof value === 'object') {
    Object.freeze(value)
    for (const item of Object.values(value as Record<string, unknown>)) deepFreeze(item)
  }
  return value
}

async function abortable<T>(value: T | Promise<T>, signal: AbortSignal): Promise<T> {
  throwIfAborted(signal)
  let onAbort: (() => void) | undefined
  const cancelled = new Promise<never>((_resolve, reject) => {
    onAbort = () => reject(signal.reason ?? new Error('decision cancelled'))
    signal.addEventListener('abort', onAbort, { once: true })
  })
  try {
    return await Promise.race([Promise.resolve(value), cancelled])
  } finally {
    if (onAbort) signal.removeEventListener('abort', onAbort)
  }
}

function throwIfAborted(signal: AbortSignal): void {
  if (signal.aborted) throw signal.reason ?? new Error('operation cancelled')
}

function assertUniqueHookIds(hooks: readonly Readonly<{ id: string }>[]): void {
  const seen = new Set<string>()
  for (const hook of hooks) {
    if (!hook.id.trim()) throw new Error('hook id must not be empty')
    if (seen.has(hook.id)) throw new Error(`duplicate hook id: ${hook.id}`)
    seen.add(hook.id)
  }
}

File: mini-agent-harness/python/extension_decision.py
from __future__ import annotations

import inspect
from collections.abc import Awaitable, Callable, Mapping, Sequence
from copy import deepcopy
from dataclasses import dataclass
from types import MappingProxyType
from typing import Literal, Protocol

DecisionBehavior = Literal["allow", "ask", "deny"]
EvidenceBehavior = DecisionBehavior | Literal["continue", "stop"]


class CancellationView(Protocol):
    @property
    def cancelled(self) -> bool: ...
    def throw_if_cancelled(self) -> None: ...


class PermissionView(Protocol):
    allowed: bool
    reason: str


@dataclass(frozen=True)
class DecisionEvidence:
    stage: Literal["pre-hook", "policy", "resolver", "post-hook", "abort"]
    source_id: str
    behavior: EvidenceBehavior
    revision: int


@dataclass(frozen=True)
class DecisionContext:
    call_id: str
    tool_name: str
    revision: int
    input: Mapping[str, object]


@dataclass(frozen=True)
class PreHookResult:
    behavior: DecisionBehavior | None = None
    updated_input: Mapping[str, object] | None = None


@dataclass(frozen=True)
class Resolution:
    allowed: bool
    reason: str
    updated_input: Mapping[str, object] | None = None


@dataclass(frozen=True)
class PreparedDecision:
    context: DecisionContext
    decision: Resolution
    evidence: tuple[DecisionEvidence, ...]


@dataclass(frozen=True)
class PostDecision:
    continue_conversation: bool
    evidence: tuple[DecisionEvidence, ...]


PreHook = Callable[
    [DecisionContext, CancellationView],
    PreHookResult | Awaitable[PreHookResult],
]
PostHook = Callable[
    [DecisionContext, bool, CancellationView],
    bool | Awaitable[bool],
]
AskResolver = Callable[
    [DecisionContext, tuple[DecisionEvidence, ...], CancellationView],
    PermissionView | Awaitable[PermissionView],
]


class ExtensionDecisionPipeline:
    def __init__(
        self,
        *,
        pre_hooks: Sequence[tuple[str, PreHook]] = (),
        post_hooks: Sequence[tuple[str, PostHook]] = (),
        ask_resolver: AskResolver | None = None,
    ) -> None:
        ids = [hook_id for hook_id, _hook in (*pre_hooks, *post_hooks)]
        if any(not hook_id.strip() for hook_id in ids):
            raise ValueError("hook id must not be empty")
        if len(ids) != len(set(ids)):
            raise ValueError("hook ids must be unique")
        self._pre_hooks = tuple(pre_hooks)
        self._post_hooks = tuple(post_hooks)
        self._ask_resolver = ask_resolver

    async def prepare(
        self,
        *,
        call_id: str,
        tool_name: str,
        tool_input: Mapping[str, object],
        validate_schema: Callable[[Mapping[str, object]], None],
        validate_semantics: Callable[
            [Mapping[str, object]], object | Awaitable[object]
        ] | None,
        decide_policy: Callable[
            [Mapping[str, object]], PermissionView | Awaitable[PermissionView]
        ],
        signal: CancellationView,
    ) -> PreparedDecision:
        signal.throw_if_cancelled()
        context = _context(call_id, tool_name, 0, tool_input)
        evidence: list[DecisionEvidence] = []
        proposals: list[DecisionBehavior] = []
        await _validate(context.input, validate_schema, validate_semantics)

        for hook_id, hook in self._pre_hooks:
            signal.throw_if_cancelled()
            result = await _maybe_await(hook(context, signal))
            if result.updated_input is not None:
                context = _context(
                    call_id, tool_name, context.revision + 1, result.updated_input
                )
                await _validate(context.input, validate_schema, validate_semantics)
            if result.behavior is not None:
                proposals.append(result.behavior)
                evidence.append(
                    DecisionEvidence(
                        "pre-hook", hook_id, result.behavior, context.revision
                    )
                )

        signal.throw_if_cancelled()
        policy = await _maybe_await(decide_policy(context.input))
        signal.throw_if_cancelled()
        evidence.append(
            DecisionEvidence(
                "policy",
                "permission-gate",
                "allow" if policy.allowed else "deny",
                context.revision,
            )
        )
        if not policy.allowed or "deny" in proposals:
            reason = policy.reason if not policy.allowed else "pre-hook-deny"
            return PreparedDecision(
                context, Resolution(False, reason), tuple(evidence)
            )

        if "ask" in proposals:
            if self._ask_resolver is None:
                evidence.append(
                    DecisionEvidence(
                        "resolver", "missing-resolver", "deny", context.revision
                    )
                )
                return PreparedDecision(
                    context,
                    Resolution(False, "approval-required-without-resolver"),
                    tuple(evidence),
                )
            try:
                resolved = await _maybe_await(
                    self._ask_resolver(context, tuple(evidence), signal)
                )
                signal.throw_if_cancelled()
            except Exception:
                if signal.cancelled:
                    raise
                evidence.append(
                    DecisionEvidence(
                        "resolver", "resolver-error", "deny", context.revision
                    )
                )
                return PreparedDecision(
                    context,
                    Resolution(False, "approval-resolver-failed"),
                    tuple(evidence),
                )
            evidence.append(
                DecisionEvidence(
                    "resolver",
                    "ask-resolver",
                    "allow" if resolved.allowed else "deny",
                    context.revision,
                )
            )
            if not resolved.allowed:
                return PreparedDecision(
                    context,
                    Resolution(False, resolved.reason),
                    tuple(evidence),
                )
            updated_input = getattr(resolved, "updated_input", None)
            if updated_input is not None:
                context = _context(
                    call_id, tool_name, context.revision + 1, updated_input
                )
                await _validate(
                    context.input, validate_schema, validate_semantics
                )
                signal.throw_if_cancelled()
                rewritten_policy = await _maybe_await(
                    decide_policy(context.input)
                )
                signal.throw_if_cancelled()
                evidence.append(
                    DecisionEvidence(
                        "policy",
                        "permission-gate-after-resolver-rewrite",
                        "allow" if rewritten_policy.allowed else "deny",
                        context.revision,
                    )
                )
                if not rewritten_policy.allowed:
                    return PreparedDecision(
                        context,
                        Resolution(False, rewritten_policy.reason),
                        tuple(evidence),
                    )

        signal.throw_if_cancelled()
        return PreparedDecision(
            context, Resolution(True, policy.reason), tuple(evidence)
        )

    async def after(
        self,
        prepared: PreparedDecision,
        *,
        succeeded: bool,
        signal: CancellationView,
    ) -> PostDecision:
        evidence = list(prepared.evidence)
        continue_conversation = True
        for hook_id, hook in self._post_hooks:
            block = bool(
                await _maybe_await(hook(prepared.context, succeeded, signal))
            )
            if block:
                continue_conversation = False
            evidence.append(
                DecisionEvidence(
                    "post-hook",
                    hook_id,
                    "stop" if block else "continue",
                    prepared.context.revision,
                )
            )
        return PostDecision(continue_conversation, tuple(evidence))


async def _validate(
    tool_input: Mapping[str, object],
    schema: Callable[[Mapping[str, object]], None],
    semantics: Callable[[Mapping[str, object]], object | Awaitable[object]] | None,
) -> None:
    schema(tool_input)
    if semantics is not None:
        await _maybe_await(semantics(tool_input))


async def _maybe_await(value):
    return await value if inspect.isawaitable(value) else value


def _context(
    call_id: str,
    tool_name: str,
    revision: int,
    tool_input: Mapping[str, object],
) -> DecisionContext:
    return DecisionContext(
        call_id,
        tool_name,
        revision,
        MappingProxyType(deepcopy(dict(tool_input))),
    )
